src/codeeditor/editor.ts

```diff
@@ -0,0 +1,637 @@
+     * Return `true` to prevent the default handling of the event by the
+     * editor.
+     */
+    onKeyDown: KeydownHandler;
+
+    /**
+     * Brings browser focus to this editor text.
+     */
+    focus(): void;
+
+    /**
+     * Repaint editor. 
+     */
+    refresh(): void;
+
+    /**
+     * Test whether the editor has keyboard focus.
+     */
+    hasFocus(): boolean;
+
+    /**
+     * Set the size of the editor in pixels.
+     */
+    setSize(width: number, height: number): void;
+
+    /**
+     * Scroll the given cursor position into view.
+     */
+    scrollIntoView(pos: IPosition, margin?: number): void;
+
+    /**
+     * Get the window coordinates given a cursor position.
+     */
+    getCoords(position: IPosition): ICoords;
+  }
+
+  /**
+   * The options used to initialize an editor.
+   */
+  export
+  class IOptions {
+    /**
+     * Whether line numbers should be displayed. Defaults to false.
+     */
+    lineNumbers?: boolean;
+
+    /**
+     * Set to false for horizontal scrolling. Defaults to true.
+     */
+    wordWrap?: boolean;
+
+    /**
+     * Whether the editor is read-only.  Defaults to false.
+     */
+    readOnly?: boolean;
+  }
+
+  /**
+   * The default implementation of the code editor model.
+   */
+  export
+  class Model implements IModel {
+    /**
+     * A signal emitted when a content of the model changed.
+     */
+    valueChanged: ISignal<this, IChangedArgs<string>>;
+
+    /**
+     * A signal emitted when a mimetype changes.
+     */
+    mimeTypeChanged: ISignal<this, IChangedArgs<string>>;
+
+    /**
+     * Whether the model is disposed.
+     */
+    get isDisposed(): boolean {
+      return this._isDisposed;
+    }
+
+    /**
+     * Dipose of the resources used by the model.
+     */
+    dispose(): void {
+      if (this._isDisposed) {
+        return;
+      }
+      this._isDisposed = true;
+      clearSignalData(this);
+    }
+
+    /**
+     * A mime type of the model.
+     */
+    get mimeType(): string {
+      return this._mimetype;
+    }
+    set mimeType(newValue: string) {
+      let oldValue = this._mimetype;
+      if (oldValue === newValue) {
+        return;
+      }
+      this._mimetype = newValue;
+      this.mimeTypeChanged.emit({
+        name: 'mimeType',
+        oldValue,
+        newValue
+      });
+    }
+
+    /**
+     * The text stored in the model.
+     */
+    get value(): string {
+      return this._value;
+    }
+    set value(newValue: string) {
+      let oldValue = this._value;
+      if (oldValue === newValue) {
+        return;
+      }
+      this._value = newValue;
+      this.valueChanged.emit({
+        name: 'value',
+        oldValue,
+        newValue
+      });
+    }
+
+    /**
+     * Get the selections for the model.
+     */
+    get selections(): ObservableVector<ITextSelection> {
+      return this._selections;
+    }
+
+    /**
+     * Returns the primary cursor position.
+     */
+    getCursorPosition(): IPosition {
```

Comment:
Maybe we better to add auxiliary functions to CodeEditor namespace instead of reimplementing them each concrete editor. I mean functions which are useful shortcuts and based on CodeEditor api.